packit_service/worker/handlers/distgit.py

```diff
@@ -644,46 +651,107 @@ def __init__(
 
     def run(self) -> TaskResults:
-        try:
-            branch = None
-            for branch in self.get_branches():
-                self.packit_api.build(
-                    dist_git_branch=branch,
+        errors = {}
+        group = self._get_or_create_koji_group_model()
+        for koji_build_model in group.grouped_targets:
+            branch = koji_build_model.target
+
+            # skip submitting build for a branch if we already did that (even if it failed)
+            if koji_build_model.status not in ["queued", "pending", "retry"]:
+                logger.debug(
+                    f"Skipping downstream Koji build for branch {branch} "
+                    f"that was already processed."
+                )
+                continue
+
+            logger.debug(f"Running downstream Koji build for {branch}")
+            koji_build_model.set_status("pending")
+
+            try:
+                stdout = self.packit_api.build(
+                    dist_git_branch=koji_build_model.target,
                     scratch=self.job_config.scratch,
                     nowait=True,
                     from_upstream=False,
                 )
-        except PackitException as ex:
-            if self.celery_task and not self.celery_task.is_last_try():
-                logger.debug(
-                    "Celery task will be retried. User will not be notified about the failure."
-                )
-                raise ex
-
-            self.report_in_issue_repository(branch, ex)
-            raise ex
+                task_id, web_url = get_koji_task_id_and_url_from_stdout(stdout)
+                koji_build_model.set_task_id(str(task_id))
+                koji_build_model.set_web_url(web_url)
+            except PackitException as ex:
+                if self.celery_task and not self.celery_task.is_last_try():
+                    kargs = self.celery_task.task.request.kwargs.copy()
+                    kargs["koji_group_model_id"] = group.id
+                    for model in group.grouped_targets:
+                        model.set_status("retry")
+
+                    logger.debug(
+                        "Celery task will be retried. User will not be notified about the failure."
+                    )
+                    retry_backoff = int(
+                        getenv("CELERY_RETRY_BACKOFF", DEFAULT_RETRY_BACKOFF)
+                    )
+                    delay = retry_backoff * 2**self.celery_task.retries
+                    self.celery_task.task.retry(exc=ex, countdown=delay, kwargs=kargs)
+                    return TaskResults(
+                        success=True,
+                        details={
+                            "msg": f"There was an error: {ex}. Task will be retried."
+                        },
+                    )
+
+                errors[branch] = str(ex)
+                # TODO store the error in the DB as well??
+                koji_build_model.set_status("error")
```

Comment:
I think this will be handy to have -- we can show it on the dashboard e.g.

Reply:
implemented, PTAL :pray: